support pid.available of DefaultEvictionHard

Cweiping opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:
`PIDPressure` ahead of time to avoid other pods of the node being unable to use PID.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/98206

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
action required: Add default Linux's eviction-hard to pid.available=10%, not configured pid.available Users of will be affected. 
If the available pid of the node (Linux) is lower than the 10% threshold, some pods will be evicted. 
The cluster administrator can refer to check whether the node workload reaches the threshold.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
1. PID based eviction: https://kubernetes.io/docs/concepts/policy/pid-limiting/#pid-based-eviction
2. Eviction Signals: https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-signals 
```
/release-note-none
/kind bug
/sig node

Commit:
add  pid.available DefaultEvictionHard

Signed-off-by: Weiping Cai <weiping.cai@daocloud.io>

Cweiping commented:
/release-note-none
/test pull-kubernetes-bazel-test

kikisdeliveryservice commented:
I think this would need a release note - @ehashman wdyt?

/triage accepted

SergeyKanzhelev commented on pkg/kubelet/apis/config/v1beta1/defaults_linux.go:
```diff
@@ -24,4 +24,5 @@ var DefaultEvictionHard = map[string]string{
 	"nodefs.available":  "10%",
 	"nodefs.inodesFree": "5%",
 	"imagefs.available": "15%",
+	"pid.available":     "10%",
```
docs will need to be updated: https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#hard-eviction-thresholds

Cweiping replied:
Done. 
Fix doc:https://github.com/kubernetes/website/pull/26255

SergeyKanzhelev commented:
/lgtm

need release notes and issue for the documentation update. Unhold when done, please

/hold

Cweiping commented:
> /lgtm
> 
> need release notes and issue for the documentation update. Unhold when done, please
> 
> /hold

Done. Awaiting your review again.

ehashman commented:
/hold cancel

I think this just needs approval now.

ehashman commented:
/priority important-longterm

mrunalp commented:
/approve

ehashman commented:
/sig api-machinery

Cweiping commented:
/assign @smarterclayton
Looking forward to your review.

SergeyKanzhelev commented:
ping @kubernetes/api-approvers

smarterclayton commented:
So this is a fairly significant operational change to a cluster, right?  After upgrade to a version with this code, workloads that were running within 10% of the max would start being evicted?  How would an administrator know that ahead of time?  What preparatory actions should an admin take before the upgrade?  These are questions I'd expect to see answered in both release notes and "pre-upgrade procedures".  Do we have an estimate of the percentage of clusters in the world this would impact after upgrade (i.e. how many people are in this state now as part of normal operations)?

Cweiping commented:
> So this is a fairly significant operational change to a cluster, right? After upgrade to a version with this code, workloads that were running within 10% of the max would start being evicted? How would an administrator know that ahead of time? What preparatory actions should an admin take before the upgrade? These are questions I'd expect to see answered in both release notes and "pre-upgrade procedures". Do we have an estimate of the percentage of clusters in the world this would impact after upgrade (i.e. how many people are in this state now as part of normal operations)?

Not configured `pid.available` Users of will be affected, others will be ignored.
If the [available pid of the node (Linux)](https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#eviction-signals) is lower than the `10%` threshold, some pods will be evicted.

The cluster administrator can refer to check whether the node workload reaches the threshold.
```shell
$ curl -s -k --key /etc/certs/kubelet-client.key --cert /etc/certs/kubelet-client.crt  https://10.6.125.8:10250/stats/summary|jq .node.rlimit.curproc/.node.rlimit.maxpid

0.0066375732421875
```
Before upgrading, the load of such nodes should be reduced as much as possible, so as to avoid driving behavior.

If cluster enhancement has been done for kubelet or `kernel.pid_max` is large enough, it will not be affected.
I found that `AliCloud`, `GCP` and `Daocloud` manufacturers have all strengthened.
The users who are affected are those who have not strengthened the reliability of their nodes. Therefore, their node reliability will be improved.

SergeyKanzhelev commented:
@Cweiping I think what @smarterclayton is suggesting is not to explain how to do it, but instead make sure we made enough preparatory actions to avoid breaking customers on upgrade.

Actions might include:

- [ ] Update release notes on this PR with an additional [action required](https://www.kubernetes.dev/docs/guide/release-notes/#applying-a-release-note:~:text=For%20pull%20requests%20that%20require%20additional,(case%20insensitive)%20in%20the%20release%20note%3A) note explaining what have changed.
- [ ] Documentation is updated: https://github.com/kubernetes/website/pull/26255. Is this PR closed intentionally?
- [ ] Think of more places this additional documentation you listed may be beneficial to be added to make sure the upgrade instructions are clear. Maybe this page https://kubernetes.io/docs/concepts/policy/pid-limiting/#m-docs-concepts-policy-pid-limiting . Also that page has a [note about pid_max](https://kubernetes.io/docs/concepts/policy/pid-limiting/#m-docs-concepts-policy-pid-limiting:~:text=limit%20to%20a%20low%20default%2C%20such%20as%2032768) that also may be helpful to mention with the introduction of this feature.

Cweiping commented:
Thank @SergeyKanzhelev and @smarterclayton review. I have updated the release note and website documents. I have updated the related documents of this feature. Is there any todo. Do you want to upgrade this PR to a feature to attract users' attention?

Associated Doc PR: https://github.com/kubernetes/website/pull/26261

ehashman commented:
@Cweiping please update your commits to pass the CLA. Are you planning on still working on this PR/issue?

liggitt commented:
this closed due to inactivity, but in case it gets picked back up, I strongly second @smarterclayton's comment at https://github.com/kubernetes/kubernetes/pull/98207#issuecomment-784350321

having the control available seems useful, but defaulting it more restrictive than current behavior does not seem generally safe. A change like this deserves a tracking KEP to surface action-required items and upgrade considerations to consumers of a new kubernetes release.

What would it look like to default this to something much closer to the edge (e.g. 1%) or not have it active by default (0%)?
